### PR TITLE
Generate initial package for Scheduled Updates

### PR DESCRIPTION
**Pull request by obenland**
This package is meant to be included with wpcomsh for WoA sites and separate from `jetpack-mu-wpcom` in case Jetpack decides to offer this functionality to self-hosted sites in the future.

## Proposed changes:

* Generate new package structure for Scheduled Updates feature.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

p1HpG7-rc9-p2#comment-69956

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* TBD

### CODE REVIEW

**Comment by obenland** [comment]
For the most part I modeled this PR after #29627, I hope this comes reasonably close to what you'd expect from a new package?

Would you suggest we load this package on WoA sites in the same way we currently include `jetpack-mu-wpcom`? 1241-gh-Automattic/wpcomsh

File: projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php
Comment (anomiex) [inline projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php]: Why is the permission check commented out? Here and elsewhere too.

File: projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php
Comment (anomiex) [inline projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php]: It looks like all the places you use `serialize()` are part of `md5( serialize() )`. You might make a static utility function that does that, with one `phpcs:ignore` that can be clearly seen to be ok, rather than scarily disabling the rule for the whole file.

File: projects/packages/scheduled-updates/phpunit.xml.dist
Comment (anomiex) [inline projects/packages/scheduled-updates/phpunit.xml.dist]: > `phpVersion="7.2.0"`

Doesn't seem like a good idea to not run the tests in 7.0. Better would be to use `set_up_before_class()` instead of `setUpBeforeClass()` (just like you already use `set_up()` instead of `setUp()`), which seems to be the only thing stopping you.

File: projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php
Comment (anomiex) [inline projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php]: Re `PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations`, see the comment in `phpunit.xml.dist`.

Re `WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize`, it may be worth doing like https://github.com/Automattic/jetpack/blob/77c6d8c8fbd58c3526ecf28e94f4378a4b84cdf9/projects/packages/forms/tests/php/.phpcs.dir.xml#L1-L4

**Review by anomiex — COMMENTED** [review]
> This package is meant to be included with wpcomsh for WoA sites and separate from jetpack-mu-wpcom in case Jetpack decides to offer this functionality to self-hosted sites in the future.

I'm not sure about that logic. Would it be that hard to just `git mv` the relevant code if it's decided to include this in Jetpack-the-plugin?

> Would you suggest we load this package on WoA sites in the same way we currently include `jetpack-mu-wpcom`? https://github.com/Automattic/wpcomsh/issues/1241

If you're going to go with a package here, the `jetpack-mu-wpcom` package itself should depend on the package and do any necessary initialization. If you try to do like Automattic/wpcomsh#1241, you'd also have to do similarly in the `jetpack-mu-wpcom-plugin` for Simple. Why duplicate the work?

File: projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php
Comment (obenland) [reply projects/packages/scheduled-updates/src/class-wpcom-rest-api-v2-endpoint-update-schedules.php]: The underlying WPCOM_Feature is not part of any plans yet since it's not launched, so this would always return false and had to be manually disabled during testing.

**Comment by obenland** [comment]
Thanks so much for your review!

> I'm not sure about that logic. Would it be that hard to just git mv the relevant code if it's decided to include this in Jetpack-the-plugin?

I'll do whatever y'all think is best. See p1HpG7-rc9-p2#comment-69956

> If you're going to go with a package here, the jetpack-mu-wpcom package itself should depend on the package and do any necessary initialization. If you try to do like https://github.com/Automattic/wpcomsh/pull/1241, you'd also have to do similarly in the jetpack-mu-wpcom-plugin for Simple. Why duplicate the work?

The functionality will always be disabled on Simple Sites since they don't update plugins/themes. So this will only run on WoA sites, and potentially Jetpack sites in the future. https://github.com/Automattic/jetpack/pull/35650 is an exploration in `jetpack-mu-wpcom`, in case it's determined that would be the way to go.

**Comment by anomiex** [comment]
> I'll do whatever y'all think is best. See p1HpG7-rc9-p2#comment-69956

That "and other standalone plugins" is being considered too pushes me to agreeing that a package makes sense. Although, depending on how likely those plans are, YAGNI may apply.

Anyway, I don't mean this to be a blocker. I just wanted to ask the question to ensure the options were considered. 🙂

> The functionality will always be disabled on Simple Sites since they don't update plugins/themes. So this will only run on WoA sites, and potentially Jetpack sites in the future. https://github.com/Automattic/jetpack/pull/35650 is an exploration in jetpack-mu-wpcom, in case it's determined that would be the way to go.

Interesting, that does make a difference in the decision making. Personally though I still think that having the jetpack-mu-wpcom package handle this as a sub-package still makes the most sense. The jetpack-mu-wpcom-plugin is really intended to be a thin wrapper around the package, existing mainly because parts of our other tooling (e.g. Jetpack Beta Tester and `jetpack-downloader`) are geared around plugins rather than packages. Making it a wrapper around more packages seems like it may be more confusing than helpful in the long run.

**Review by anomiex — COMMENTED** [review]
Anyway, from a structural perspective this looks good to me now. I'll leave deeper review and testing of the actual code to the team responsible. 🙂

**Comment by obenland** [comment]
> I still think that having the jetpack-mu-wpcom package handle this as a sub-package still makes the most sense.

If we went that route, would we be able to test PRs for this package with Jetpack Beta?

What would merges/deployments look like?

**Comment by anomiex** [comment]
> If we went that route, would we be able to test PRs for this package with Jetpack Beta?

Yes, for WoA testing it should work exactly the same as for testing changes directly in the jetpack-mu-wpcom package, with no additional setup.

The other approach would need extra setup, in that you'd need to make sure that however wpcomsh loads the additional package will do for this package the same thing it does to load the jetpack-mu-wpcom package out of the Beta-provided plugin.

> What would merges/deployments look like?

We may need to experiment with this one to make sure of the right instructions. But it should be like this:

* You'd merge your changes to the package in the monorepo same as you would now and would with the other method.
* The release process (PCYsg-Osp-p2#new-version) will cut a new version of the scheduled-updates package, even if the jetpack-mu-wpcom package doesn't have any changes. Nothing new should be needed there either.
* When it comes to the part about updating wpcomsh (PCYsg-Osp-p2#woa-deployment), we'll need to make sure that the version of the scheduled-updates package gets updated in composer.lock even if the jetpack-mu-wpcom package itself doesn't have any changes. I'm pretty sure that updating those instructions to say to do `composer update automattic/jetpack-mu-wpcom automattic/scheduled-updates` should be all that's needed if anything. Which you'd have to do with the other method too (in addition to also updating composer.json).
* The rest of the process of releasing wpcomsh will be the same as it is now and with the other method.

**Comment by obenland** [comment]
Thanks so much for your feedback and guidance on this. Is [`6cfb019` (#35796)](https://github.com/Automattic/jetpack/pull/35796/commits/6cfb019f60851a4c0dba2c2bc03572069d318e3c) going in the direction you were hoping for?

File: projects/plugins/mu-wpcom-plugin/composer.json
Comment (anomiex) [inline projects/plugins/mu-wpcom-plugin/composer.json]: This should not be needed, the addition in `projects/packages/jetpack-mu-wpcom/composer.json` should already be bringing it in.

You can test that easily enough by reverting all the changes in `projects/plugins/mu-wpcom-plugin/`, then run `tools/fixup-project-versions.sh` and see whether it updates composer.lock with the same changes (except for the hash on line 7) as shown below.

**Review by anomiex — COMMENTED** [review]
> Thanks so much for your feedback and guidance on this. Is https://github.com/Automattic/jetpack/pull/35796/commits/6cfb019f60851a4c0dba2c2bc03572069d318e3c going in the direction you were hoping for?

You're welcome! Looks good to me other than one thing mentioned inline.

File: projects/plugins/mu-wpcom-plugin/composer.json
Comment (obenland) [reply projects/plugins/mu-wpcom-plugin/composer.json]: With the changes to `mu-wpcom-plugin` reverted, I'm unable to follow the WoA testing steps in PCYsg-Osp-p2/#woa-testing. The Scheduled Updates package doesn't end up in `vendors`. Am I missing a step there?

Comment (anomiex) [reply projects/plugins/mu-wpcom-plugin/composer.json]: Did you run `tools/fixup-project-versions.sh` to re-update the composer.lock after reverting the changes in the plugin directory?

Other than that, I don't know. It seems to work for me locally.

Comment (obenland) [reply projects/plugins/mu-wpcom-plugin/composer.json]: That's what I was missing, thanks. It still updated the version for me and I thought we wanted to avoid that, so I reverted that, too.

Comment (anomiex) [reply projects/plugins/mu-wpcom-plugin/composer.json]: Sorry I wasn't clear. Yeah, it's mainly this one line in composer.json we wanted to leave out. 🙂

**Review by anomiex — APPROVED** [review]
Thanks for the additional changes!

I'll hit Approve with the note that I haven't reviewed the actual code being added in the package, so if someone hasn't done that yet you may want to wait for a second approval. 🙂

**Review by zaerl — APPROVED** [review]
LGTM!

**Pull request merged**
